### PR TITLE
attempt to keep active connections open with all bootnodes during kee…

### PR DESCRIPTION
**Pull request by hexoscott**

### CODE REVIEW

File: network/server.go [hunk 1]
```diff
@@ -327,26 +329,61 @@ func (s *Server) setupBootnodes() error {
...
+// if the active peer count is lesser than the specified limit. And will attempt
+// to ensure that all bootnodes are connected to
 func (s *Server) keepAliveMinimumPeerConnections() {
+	bootNodes := s.bootnodes.getBootnodes()
+	ourId := s.host.ID()
+
+	// remove ourselves from the list of bootNodes if present
+	for i, bootNode := range bootNodes {
+		if bootNode.ID == ourId {
+			bootNodes = append(bootNodes[:i], bootNodes[i+1:]...)
```
Comment (Stefan-Ethernal) [inline network/server.go:341]: Double check this, but not necessary looping through the entire bootNodes slice if we find the "local node" among the bootnodes:

```suggestion
		bootNodes = append(bootNodes[:i], bootNodes[i+1:]...)
        break
```

File: network/server.go [hunk 2]
```diff
@@ -327,29 +329,63 @@ func (s *Server) setupBootnodes() error {
...
-		if s.numPeers() < MinimumPeerConnections {
-			if s.config.NoDiscover || !s.bootnodes.hasBootnodes() {
-				// dial unconnected peer
-				randPeer := s.GetRandomPeer()
-				if randPeer != nil && !s.IsConnected(*randPeer) {
-					s.addToDialQueue(s.GetPeerInfo(*randPeer), common.PriorityRandomDial)
+		currentPeers := s.numPeers()
+
+		// if we're in a setup where we just want random peering
+		if currentPeers < MinimumPeerConnections && (s.config.NoDiscover || !s.bootnodes.hasBootnodes()) {
```
Comment (Stefan-Ethernal) [inline network/server.go]: Nitpick, doesn't add to verbosity of the code I'd say:

```suggestion
		// if we're in a setup where we just want random peering
		if s.numPeers() < MinimumPeerConnections && (s.config.NoDiscover || !s.bootnodes.hasBootnodes()) {
```

File: network/server.go [hunk 3]
```diff
@@ -327,29 +329,63 @@ func (s *Server) setupBootnodes() error {
...
+			randPeer := s.GetRandomPeer()
+			s.logger.Info("[serverKeepAlive] adding random peer", "id", randPeer.String())
+			if randPeer != nil && !s.IsConnected(*randPeer) {
+				s.addToDialQueue(s.GetPeerInfo(*randPeer), common.PriorityRandomDial)
+			}
+		}
+
+		// ensure we attempt to connect to all of our bootNodes
+		peers := s.Peers()
+		var added []string
```
Comment (Stefan-Ethernal) [inline network/server.go]: Optional change:
```suggestion
		added := make([]string, 0, len(bootNodes))
```

**Review by Stefan-Ethernal — APPROVED** [review]
I suppose it should be safe to make this change.

File: network/server.go (hunk 3 above)
Comment (hexoscott) [reply network/server.go]: I opted for using `MaximumDialsPerKeepAlive` as the capacity as we won't ever add more than this.

**Comment by hexoscott** [comment]
All points actioned, thank you @Stefan-Ethernal for the eyes on this one.  I'll get these changes out to testnet tomorrow morning.

**Pull request merged**